core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp

```diff
@@ -200,6 +200,9 @@ struct GraphImpl<Kokkos::Cuda> {
         m_execution_space, _graph_node_kernel_ctor_tag{},
         aggregate_kernel_impl_t{});
   }
+
+  cudaGraph_t& get_cuda_graph() { return m_graph; }
+  cudaGraphExec_t& get_cuda_graph_exec() { return m_graph_exec; }
```

Comment:
Why did you change and now return a reference?  AFAIK `cudaGraph[Exec]_t` are handle types, some typedef for a pointer to an implementation-defined struct and they are intended to be passed by value.

Reply:
That said I was about to ask about the const-qualifier in the previous version.

Reply:
### `get_cuda_graph`

I think that if we want to allow interoperability in both direction, it has to be
```c++
cudaGraph_t& get_cuda_graph() { return m_graph; }
```
(*i.e.* by reference).

Because if I have an existing `cudaGraph_t`, and want to "set it" to my `Kokkos::Graph`, there is no other way I guess. Or maybe you'd want a `set_cuda`graph` ? Up to you :smile:

### `get_cuda_graph_exec`

For this one, one of the primary reasons to handle `cudaGraphExec_t` "manually" would be when one wants to *instantiate the graph with parameters* (something the `Kokkos` API does not allow us to do for now).

The signature of one of the overloads of `cudaGraphInstantiate` from the `Cuda` documentation:
```c++
__host__ ​cudaError_t cudaGraphInstantiateWithParams ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, cudaGraphInstantiateParams* instantiateParams );
```

Reply:
OK so that was your way of addressing https://github.com/kokkos/kokkos/pull/6904#pullrequestreview-1972501757

I am not a big fan of the getter returning a reference.  I would have preferred if we could find a way to set the root node at construction and not allow anyone to change the graph from underneath Kokkos otherwise.  This makes it possible to break invariances such as mismatch between the graph and the "executable".

Regarding the setter for the graph executable.  Do you actually need it?  Are you not able to set properties from the handle when it is returned by value?

Reply:
#### Graph handle

For the `cudaGraph_t` object, I guess it's enough for now to be able to retrieve it *by value*. I suggest we use this PR to add just the regualr getter  as follows:
```c++
cudaGraph_t get_cuda_graph() const { return m_graph; }
```
This would already enable interoperability "from `Kokkos` to outside". And we could discuss in another issue how to best enable interoperability "from outside to `Kokkos`", *i.e.* how to allow construction of a `Kokkos::Graph` from an existing `cudaGraph_t` handle.

#### Executable graph handle

The function used by `Kokkos` with the error node and logs has been "removed" in `CUDA 12` (https://developer.nvidia.com/blog/cuda-toolkit-12-0-released-for-general-availability/, still in https://docs.nvidia.com/cuda/archive/11.1.0/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH but not anymore in https://docs.nvidia.com/cuda/archive/12.4.0/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH).

I looked in my include files (I'm on `Cuda 12.3.2`):
```c++
static __inline__ __host__ cudaError_t cudaGraphInstantiate( cudaGraphExec_t *pGraphExec,  cudaGraph_t graph,  cudaGraphNode_t *pErrorNode,  char *pLogBuffer,  size_t bufferSize){
  (void)pErrorNode;
  (void)pLogBuffer;
  (void)bufferSize;
  return ::cudaGraphInstantiate(pGraphExec, graph, 0);
}
```
In the new `Cuda` versions, these are the API functions that can be used to instantiate the graph:
```c++
__host__ ​cudaError_t cudaGraphInstantiate ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, unsigned long long flags = 0 )

__host__ ​cudaError_t cudaGraphInstantiateWithFlags ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, unsigned long long flags = 0 )

__host__ ​cudaError_t cudaGraphInstantiateWithParams ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, cudaGraphInstantiateParams* instantiateParams ) 
```

So we could now pass the parameters through `Kokkos::Impl::instantiate` more easily. I pushed the changes just for you to see. Note that we might enter into some `CUDA` API versioning issues (the `CUDA` graph API changed in `CUDA 12`).